include/picongpu/particles/Manipulate.hpp

```diff
@@ -20,21 +20,46 @@
 #pragma once
 
 #include "picongpu/simulation_defines.hpp"
-#include "picongpu/particles/filter/filter.def"
 #include "picongpu/particles/manipulators/manipulators.def"
+#include "picongpu/particles/filter/filter.def"
 
-#include <pmacc/Environment.hpp>
+#include <pmacc/particles/algorithm/CallForEach.hpp>
 #include <pmacc/meta/conversion/ToSeq.hpp>
 #include <pmacc/meta/ForEach.hpp>
 #include <pmacc/particles/meta/FindByNameOrType.hpp>
 
-#include <boost/mpl/apply.hpp>
-
 
 namespace picongpu
 {
 namespace particles
 {
+namespace detail
+{
+    /** Operator to create a filtered functor
+     */
+    template<
+        typename T_Filter,
+        typename T_Manipulator,
+        typename T_SpeciesType
+    >
+    struct MakeUnaryFilteredFunctor
+    {
+
+        using SpeciesType = pmacc::particles::meta::FindByNameOrType_t<
+            VectorAllSpecies,
+            T_SpeciesType
+        >;
+        using SpeciesFunctor = typename bmpl::apply1<
+            T_Manipulator,
+            SpeciesType
+        >::type;
+        using type = manipulators::IUnary<
```

Comment:
Maybe only `type` should be public (and `SpeciesType`, `SpeciesFunctor` private) to be even more clear what the user interface is? The same argument could be applied for `Manipulate`.